File: apps/mobile/src/lib/__mocks__/RNEthersRs.ts
import { providers, utils, Wallet } from 'ethers'

const pathFromIndex = (index: number): string => `m/44'/60'/0'/0/${index}`

const mnemonics: { [id: string]: string } = {}
const privateKeys: { [id: string]: string } = {}

export function getMnemonicIds(): Promise<string[]> {
  return Promise.resolve(Object.keys(mnemonics))
}

// returns the mnemonicId (derived address at index 0) of the imported mnemonic
export function importMnemonic(mnemonic: string): Promise<string> {
  const wallet = Wallet.fromMnemonic(mnemonic)
  mnemonics[wallet.address] = mnemonic
  return Promise.resolve(wallet.address)
}

// returns the mnemonicId (derived address at index 0) of the stored mnemonic
export function generateAndStoreMnemonic(): Promise<string> {
  const wallet = Wallet.createRandom()
  mnemonics[wallet.address] = wallet.mnemonic.phrase
  return Promise.resolve(wallet.address)
}

export function getAddressesForStoredPrivateKeys(): Promise<string[]> {
  return Promise.resolve(Object.keys(privateKeys))
}

// returns the address for the mnemonic
export function generateAddressForMnemonic(
  mnemonic: string,
  derivationIndex: number
): Promise<string> {
  const wallet = Wallet.fromMnemonic(mnemonic, pathFromIndex(derivationIndex))
  return Promise.resolve(wallet.address)
}

// returns the address of the generated key
export function generateAndStorePrivateKey(
  mnemonicId: string,
  derivationIndex: number
): Promise<string> {
  const mnemonic = mnemonics[mnemonicId]
  if (!mnemonic) {
    return Promise.reject(`No mnemonic found for ${mnemonicId}`)
  }
  const wallet = Wallet.fromMnemonic(mnemonic, pathFromIndex(derivationIndex))
  privateKeys[wallet.address] = wallet.privateKey
  return Promise.resolve(wallet.address)
}

export async function signTransactionForAddress(
  address: string,
  transaction: providers.TransactionRequest
): Promise<string> {
  const privateKey = privateKeys[address]
  if (!privateKey) {
    return Promise.reject(`No private key found for ${address}`)
  }
  const wallet = new Wallet(privateKey)
  const signature = await wallet.signTransaction(transaction)
  return signature
}

export async function signMessageForAddress(
  address: string,
  message: string | utils.Bytes
): Promise<string> {
  const privateKey = privateKeys[address]
  if (!privateKey) {
    return Promise.reject(`No private key found for ${address}`)
  }
  const wallet = new Wallet(privateKey)
  const signature = await wallet.signMessage(message)
  return signature
}

File: apps/mobile/ios/Uniswap/RNEthersRs/RnEthersRS.m
//
//  RnEthersBridge.m
//  Uniswap
//
//  Created by Connor McEwen on 10/28/21.
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(RNEthersRS, NSObject)

RCT_EXTERN_METHOD(getMnemonicIds: (RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(importMnemonic: (NSString *)mnemonic
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(removeMnemonic: (NSString *)mnemonicId
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(generateAndStoreMnemonic: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getAddressesForStoredPrivateKeys: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(generateAddressForMnemonic: (NSString *)mnemonic
                  derivationIndex: (NSInteger)index
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(generateAndStorePrivateKey: (NSString *)mnemonicId
                  derivationIndex: (NSInteger)index
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(removePrivateKey: (NSString *)address
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(signTransactionHashForAddress: (NSString *)address
                  hash: (NSString *)hash
                  chainId: NSNumber
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(signMessageForAddress: (NSString *)address
                  message: (NSString *)message
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(signHashForAddress: (NSString *)address
                  hash: (NSString *)hash
                  chainId: NSNumber
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject)

@end
